src/test/shell/bazel/external_files_test.sh

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env bash
+#
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Test the local files overlay patching functionality of external repositories.
+#
+
+# Load the test setup defined in the parent directory
+CURRENT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${CURRENT_DIR}/../integration_test_setup.sh" \
+  || { echo "integration_test_setup.sh not found!" >&2; exit 1; }
+
+
+src_only_repo_with_local_include() {
+  # Generates source files in the current directory.
+  mkdir src
+  cat > src/main.c <<'EOF'
+#include <stdio.h>
+#include "src/consts/greeting.h"
+
+int main(int argc, char **argv) {
+  printf("%s\n", GREETING);
+  return 0;
+}
+EOF
+  mkdir src/consts
+  cat > src/consts/greeting.h <<'EOF'
+#define GREETING "Hello World"
+EOF
+}
+
+build_def() {
+  # Generates build and module definition files for `src_only_repo_with_local_include` in the
+  # current directory.
+  cat > _.MODULE <<'EOF'
+module(name = "remote")
+EOF
+  mkdir src
+  cat > _.BUILD <<'EOF'
+cc_binary(
+  name = "hello",
+  srcs = ["main.c", "consts/greeting.h"],
+)
+EOF
+}
+
+test_files_repository() {
+  # Verify that overlaid files are correctly positioned.
+  WRKDIR=$(mktemp -d "${TEST_TMPDIR}/testXXXXXX")
+  cd "${WRKDIR}"
+
+  mkdir remote
+  (cd remote && src_only_repo_with_local_include)
+  tar cvf remote.tar remote
+  rm -rf remote
+
+  mkdir main
+  cd main
+  build_def
+  touch BUILD
+  cat >> $(setup_module_dot_bazel) <<EOF
+http_archive = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+http_archive(
+  name = "remote",
+  strip_prefix = "remote",
+  urls = ["file://${WRKDIR}/remote.tar"],
+  files = {
+    "MODULE.bazel": "//:_.MODULE",
```

Comment:
Can you please also add a test case to verify if the repo is re-fetched if the file content changes?

Reply:
This is implicit (symlinks are always up-to-date so no need to `watch` files), regardless I've added a test case to be sure. Useful in the event the implementation changes to a file copy in the future (use cases exist, namely for NodeJS which follows symlinks by default).

Reply:
On windows, symlink might be implemented as a copy if `--windows_enable_symlinks` isn't turned on. I think it's probably worth watching those files anyway.

Reply:
Good point, updated.